### PR TITLE
Fix #939 : Add WalkthroughActivity Landscape Support

### LINKED ISSUE #939: Lowfi Landscape WalkthroughActivity
Mocks: 
https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/2137efb6-6940-498d-8d7d-a8ea87e98410/UP-User-Profile-21
https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/7f7a12db-2dde-442c-a236-5953e36ca840/UP-User-Profile-22
https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/f3384f81-32b2-40ee-b961-b153e77ac988/UP-User-Profile-23

Add support for landscape mode and also add test cases for the same.

### PR DESCRIPTION
**Pull request by aggarwalpulkit596**
## Explanation
Fixes #939 

Mock Link: https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/2137efb6-6940-498d-8d7d-a8ea87e98410/UP-User-Profile-21
https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/7f7a12db-2dde-442c-a236-5953e36ca840/UP-User-Profile-22
https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/f3384f81-32b2-40ee-b961-b153e77ac988/UP-User-Profile-23

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by aggarwalpulkit596** [comment]
@rt4914 can you take a look at the layouts are they good enough if yes then I can proceed with the test for landscape

**Review by rt4914 — CHANGES_REQUESTED** [review]
Screenshot_1586239962

Screenshot_1586239964

This is how UI is looking. Also, now as you have submitted enough PR the correct approach would be to finish the PR from your end first including the test cases and then assign it for review. 
Approach review should not take place unless you are completely stuck on something.

**Comment by aggarwalpulkit596** [comment]
@rt4914 I will keep that in mind

**Comment by aggarwalpulkit596** [comment]
@rt4914 there only one case which is i'm not able handle which is when user press back button in that case how can i get to know the current page index

**Comment by aggarwalpulkit596** [comment]
@rt4914 maybe try out by local build you'll understand better what I'm trying to explain

**Comment by aggarwalpulkit596** [comment]
@rt4914 should I move forward with this approach now?  Only need to write some test now rest all is good i guess

**Comment by rt4914** [comment]
> @rt4914 should I move forward with this approach now? Only need to write some test now rest all is good i guess

Yeah, this approach does look correct to me.

File: app/src/main/java/org/oppia/app/walkthrough/topiclist/WalkthroughTopicListFragmentPresenter.kt [hunk 1]
```diff
@@ -38,11 +41,16 @@ class WalkthroughTopicListFragmentPresenter @Inject constructor(
...
+      2
+    } else {
+      3
+    }
+    val walkthroughLayoutManager = GridLayoutManager(activity.applicationContext, spanCount)
     walkthroughLayoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
       override fun getSpanSize(position: Int): Int {
         return if (position == 0) {
-          /* number of spaces this item should occupy = */ 2
+          /* number of spaces this item should occupy = */spanCount
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/walkthrough/topiclist/WalkthroughTopicListFragmentPresenter.kt]: Nit: `/* number of spaces this item should occupy = */ spanCount`

File: app/src/main/res/layout-land/walkthrough_final_fragment.xml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+  <data>
+
+    <variable
+      name="viewModel"
+      type="org.oppia.app.walkthrough.end.WalkthroughFinalViewModel" />
+    <variable
+      name="presenter"
+      type="org.oppia.app.walkthrough.end.WalkthroughFinalFragmentPresenter" />
+  </data>
```
Comment (rt4914) [inline app/src/main/res/layout-land/walkthrough_final_fragment.xml:15]: Formatting is incorrect. You can use https://docs.google.com/document/d/1AHIJagMfATgpbNCFmYoYaSqzsKp8JjTgPLQFXEFJdUA/edit?usp=sharing to fix that formatting.

File: app/src/main/res/layout-land/walkthrough_final_fragment.xml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        android:layout_width="180dp"
+        android:layout_height="180dp"
+        android:layout_marginStart="28dp"
+        android:src="@drawable/ic_onboarding_0"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+      <LinearLayout
+        android:id="@+id/linearLayout3"
```
Comment (rt4914) [inline app/src/main/res/layout-land/walkthrough_final_fragment.xml]: Please provide appropriate ID which is meaningful

File: app/src/main/res/layout-land/walkthrough_welcome_fragment.xml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    <variable
+      name="viewModel"
+      type="org.oppia.app.walkthrough.welcome.WalkthroughWelcomeViewModel" />
+    <variable
+      name="presenter"
+      type="org.oppia.app.walkthrough.welcome.WalkthroughWelcomeFragmentPresenter" />
+  </data>
+
+  <androidx.constraintlayout.widget.ConstraintLayout
```
Comment (rt4914) [inline app/src/main/res/layout-land/walkthrough_welcome_fragment.xml:17]: Ditto as above

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughActivityTest.kt [hunk 5]
```diff
@@ -41,6 +41,17 @@ class WalkthroughActivityTest {
     }
   }
 
+
+  @Test
+  fun testWalkthroughFragment_increaseProgress_configurationChanged_worksCorrectly() {
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughActivityTest.kt:50]: Add back button test case also, meaning:

1. Go to second fragment
2. Orientation change
3. Click on back button
4. Check the content.

**Review by rt4914 — CHANGES_REQUESTED** [review]
Changes suggested. Also please add more detailed test cases for all other child fragments too.

File: app/src/main/res/layout-land/walkthrough_final_fragment.xml (hunk 2 above)
Comment (aggarwalpulkit596) [reply app/src/main/res/layout-land/walkthrough_final_fragment.xml:15]: Thanks for this

**Review by rt4914 — CHANGES_REQUESTED** [review]
I think you missed this: https://github.com/oppia/oppia-android/pull/943#pullrequestreview-397165351

Please add more test cases for landscape mode across other fragments too.

**Comment by aggarwalpulkit596** [comment]
> I think you missed this: [#943 (review)](https://github.com/oppia/oppia-android/pull/943#pullrequestreview-397165351)
> 
> Please add more test cases for landscape mode across other fragments too.

okay i thought that only test case regarding the back button was supposed to be added 
Nm will add more tests right away.

**Comment by aggarwalpulkit596** [comment]
@rt4914 I've added some more test couldn't think of any more test.

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughWelcomeFragmentTest.kt [hunk 6]
```diff
@@ -85,6 +86,15 @@ class WalkthroughWelcomeFragmentTest {
     }
   }
 
+  @Test
+  fun testWalkthroughWelcomeFragment_checkProfileName_configurationChanged_isCorrect() {
+    launch<OnboardingActivity>(createWalkthroughActivityIntent(0)).use {
+      onView(allOf(withId(R.id.walkthrough_welcome_title_text_view), isCompletelyDisplayed())).check(matches(withText("Welcome Sean!")))
+      onView(isRoot()).perform(OrientationChangeAction.orientationLandscape())
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughWelcomeFragmentTest.kt]: Nit:
`onView(isRoot()).perform(orientationLandscape())`

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughTopicListFragmentTest.kt [hunk 7]
```diff
@@ -64,4 +65,18 @@ class WalkthroughTopicListFragmentTest {
       )
     }
   }
+
+  @Test
+  fun testWalkthroughWelcomeFragment_recyclerViewIndex1_topicCard_configurationChanged_topicNameIsCorrect() {
+    launch<WalkthroughActivity>(createWalkthroughActivityIntent(0)).use {
+      onView(withId(R.id.walkthrough_welcome_next_button)).perform(click())
+      onView(withId(R.id.walkthrough_topic_recycler_view)).perform(scrollToPosition<RecyclerView.ViewHolder>(1))
+      onView(isRoot()).perform(OrientationChangeAction.orientationLandscape())
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughTopicListFragmentTest.kt]: Nit:
`onView(isRoot()).perform(orientationLandscape())`

Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughTopicListFragmentTest.kt]: Optimise imports accordingly

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughWelcomeFragmentTest.kt (hunk 6 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughWelcomeFragmentTest.kt]: Optimise imports accordingly

**Review by rt4914 — APPROVED** [review]
LGTM, nit change suggested.

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughTopicListFragmentTest.kt (hunk 7 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughTopicListFragmentTest.kt]: Will do right away

**Comment by aggarwalpulkit596** [comment]
@rt4914 done also importing functions like the above method does affect on code performance or it's just better looking code?

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

**Comment by rt4914** [comment]
@aggarwalpulkit596 Merge this code with latest develop as it is stuck on linters test.

**Comment by aggarwalpulkit596** [comment]
@rt4914 doing in a minute

**Pull request merged**